PR: https://github.com/odoo/odoo/pull/282884

From: 78b5ef88fb0245c059449db4f2d24c91ee5317df
From: svs-odoo
Date: 2026-08-28 20:53:37

Breaking data model changes scores: del:1 + add:6, change matches:
-    module_stock_picking_batch = fields.Boolean("Batch, Wave & Cluster Transfers")
+    picking_ids = fields.One2many(
+    allowed_picking_ids = fields.One2many('stock.picking', compute='_compute_allowed_picking_ids')
+    move_ids = fields.One2many(
+    move_line_ids = fields.One2many(
+    wave_category_ids = fields.Many2many('product.category',
+    wave_location_ids = fields.Many2many('stock.location',

Total Changes: 1273

[IMP] stock,*: move picking batch into stock

*: delivery_stock_picking_batch, l10n_ar_stock, l10n_ar_stock_batch,
stock_delivery, stock_fleet, stock_picking_batch

Before this commit, the batch feature has its own module,
`stock_picking_batch`, which means if another module need to work with
badge, if that module only requires `stock`, a bridge module with
`stock_picking_batch` is needed.

The idea is to move all data (model, views, ect.) directly into `stock`
and to enable the feature with a config setting boolean field (like for
lots & serial numbers with the tracking setting for example).

Some modules were removed:
* `delivery_stock_picking_batch` (merged into `stock_delivery`);
* `l10n_ar_stock_batch` (merged into `l10_ar_stock`);
* `l10n_ro_edi_stock_batch` (merged into `l10n_ro_edi_stock`);
* `stock_picking_batch` (merged into `stock`).

task-6154300

Part-of: odoo/odoo#282884
Related: odoo/enterprise#128175
Related: odoo/upgrade#11054
Signed-off-by: William Henrotin (whe) <whe@odoo.com>

================================= pseudo patch: =================================

--- a/addons/stock/models/__init__.py
+++ b/addons/stock/models/__init__.py
@@ -23,6 +23,7 @@ from . import stock_package_history
 from . import stock_package_type
 from . import stock_package
 from . import stock_picking
+from . import stock_picking_batch
 from . import stock_picking_type
 from . import stock_storage_category
 from . import stock_scrap_reason_tag

--- a/addons/stock/models/res_config_settings.py
+++ b/addons/stock/models/res_config_settings.py
@@ -16,6 +16,8 @@ class ResConfigSettings(models.TransientModel):
         implied_group='stock.group_stock_lot_print_gs1')
     group_lot_on_delivery_slip = fields.Boolean("Display Lots & Serial Numbers on Delivery Slips",
         implied_group='stock.group_lot_on_delivery_slip', group="base.group_user,base.group_portal")
+    group_stock_picking_batch = fields.Boolean("Batch, Wave & Cluster Transfers",
+        implied_group='stock.group_stock_picking_batch')
     group_stock_tracking_lot = fields.Boolean("Packages",
         implied_group='stock.group_tracking_lot')
     group_stock_tracking_owner = fields.Boolean("Consignment",
@@ -25,7 +27,6 @@ class ResConfigSettings(models.TransientModel):
         help="Add and customize route operations to process product moves in your warehouse(s): e.g. unload > quality control > stock for incoming products, pick > pack > ship for outgoing products. \n You can also set putaway strategies on warehouse locations in order to send incoming products into specific child locations straight away (e.g. specific bins, racks).")
     group_warning_stock = fields.Boolean("Warnings for Stock", implied_group='stock.group_warning_stock')
     group_stock_sign_delivery = fields.Boolean("Signature", implied_group='stock.group_stock_sign_delivery')
-    module_stock_picking_batch = fields.Boolean("Batch, Wave & Cluster Transfers")
     module_stock_barcode = fields.Boolean("Barcode Scanner")
     module_stock_barcode_barcodelookup = fields.Boolean("Stock Barcode Database")
     stock_move_email_validation = fields.Boolean(related='company_id.stock_move_email_validation', readonly=False)

--- a/addons/stock/models/stock_move.py
+++ b/addons/stock/models/stock_move.py
@@ -929,6 +929,12 @@ Please change the quantity done or the rounding precision in your settings.""",
             self._update_orderpoints()
         if 'picking_id' in vals:
             self._set_references()
+        if self.env.user.has_group('stock.group_stock_picking_batch') and\
+           'state' in vals and vals['state'] in ('partially_available', 'assigned'):
+            for picking in self.picking_id:
+                if picking.state != 'assigned':
+                    continue
+                picking._find_auto_batch()
         return res
 
     def _update_orderpoints(self):
@@ -1032,8 +1038,7 @@ Please change the quantity done or the rounding precision in your settings.""",
         """
         self.ensure_one()
         view = self.env.ref('stock.view_stock_move_operations')
-
-        return {
+        action = {
             'name': _('Detailed Operations'),
             'type': 'ir.actions.act_window',
             'view_mode': 'form',
@@ -1047,6 +1052,11 @@ Please change the quantity done or the rounding precision in your settings.""",
                 auto_pick_move_lines=self.picked,
             ),
         }
+        if self.env.context.get('show_picking') and self.picking_id.batch_id:
+            action['name'] = self.env._('Open: Stock Move')
+            action['context']['default_picking_id'] = self.picking_id.id
+            action['context']['display_name_partner'] = True
+        return action
 
     def action_product_forecast_report(self):
         self.ensure_one()
@@ -1586,16 +1596,22 @@ Please change the quantity done or the rounding precision in your settings.""",
         return keys
 
     def _search_picking_for_assignation_domain(self):
-        domain = [
+        domain = Domain([
             ('reference_ids', 'in', self.reference_ids.ids),
             ('location_id', '=', self.location_id.id),
             ('location_dest_id', '=', (self.location_dest_id.id or self.picking_type_id.default_location_dest_id.id)),
             ('picking_type_id', '=', self.picking_type_id.id),
             ('printed', '=', False),
-            ('state', 'in', ['draft', 'confirmed', 'waiting', 'partially_available', 'assigned'])]
+            ('state', 'in', ['draft', 'confirmed', 'waiting', 'partially_available', 'assigned']),
+        ])
         if self.partner_id:
             picking_partner_id = self.env.context.get('move_picking_partner_id', self.partner_id).id
-            domain += [('partner_id', '=', picking_partner_id)]
+            domain = Domain.AND([domain, [('partner_id', '=', picking_partner_id)]])
+        if self.env.user.has_group('stock.group_stock_picking_batch'):
+            domain = Domain.AND([
+                domain,
+                ['|', ('batch_id', '=', False), ('batch_id.is_wave', '=', False)],
+            ])
         return domain
 
     def _search_picking_for_assignation(self):
@@ -1652,6 +1668,9 @@ Please change the quantity done or the rounding precision in your settings.""",
         return vals
 
     def _assign_picking_post_process(self, new=False):
+        if self.env.user.has_group('stock.group_stock_picking_batch'):
+            for picking in self.picking_id:
+                picking._find_auto_batch()
         pass
 
     def _generate_serial_move_line_commands(self, field_data, location_dest_id=False, origin_move_line=None):
@@ -2240,6 +2259,9 @@ Please change the quantity done or the rounding precision in your settings.""",
             self.picking_id._check_entire_pack()
         StockMove.browse(moves_to_redirect).move_line_ids._apply_putaway_strategy()
 
+        if self.env.user.has_group('stock.group_stock_picking_batch'):
+            self.move_line_ids._auto_wave()
+
     def _action_cancel(self):
         if any(move.state == 'done' and move.location_dest_usage != 'inventory' for move in self):
             raise UserError(_('You cannot cancel a stock move that has been set to \'Done\'. Create a return in order to reverse the moves which took place.'))
@@ -2280,6 +2302,11 @@ Please change the quantity done or the rounding precision in your settings.""",
             'move_orig_ids': [(5, 0, 0)],
             'procure_method': 'make_to_stock',
         })
+
+        for picking in self.picking_id:
+            # Remove the picking from the batch if the whole batch isn't cancelled.
+            if picking.state == 'cancel' and picking.batch_id and any(p.state != 'cancel' for p in picking.batch_id.picking_ids):
+                picking.batch_id = None
         return True
 
     def _log_cancel_activity(self):

--- a/addons/stock/models/stock_move_line.py
+++ b/addons/stock/models/stock_move_line.py
@@ -1,4 +1,3 @@
-# -*- coding: utf-8 -*-
 # Part of Odoo. See LICENSE file for full copyright and licensing details.
 
 from collections import Counter, defaultdict
@@ -23,6 +22,7 @@ class StockMoveLine(models.Model):
         check_company=True,
         index=True,
         help='The stock operation where the packing has been made')
+    batch_id = fields.Many2one(related='picking_id.batch_id')
     move_id = fields.Many2one(
         'stock.move', 'Stock Operation',
         check_company=True, index=True)
@@ -1233,3 +1233,398 @@ class StockMoveLine(models.Model):
     def _should_set_package(self):
         package_type = self.picking_id.picking_type_id
         return len(package_type) == 1 and package_type.set_package_type
+
+    # ===== Batch business methods =====
+
+    def action_open_add_to_wave(self):
+        # Can be called from the move line list view or from the 'Add to wave' wizard.
+        if 'active_wave_id' in self.env.context:
+            wave = self.env['stock.picking.batch'].browse(self.env.context.get('active_wave_id'))
+            return self._add_to_wave(wave)
+        view = self.env.ref('stock.stock_add_to_wave_form')
+        return {
+            'name': _('Add to Wave'),
+            'type': 'ir.actions.act_window',
+            'view_mode': 'form',
+            'res_model': 'stock.add.to.wave',
+            'views': [(view.id, 'form')],
+            'view_id': view.id,
+            'target': 'new',
+        }
+
+    def _add_to_wave(self, wave=False, description=False):
+        """ Detach lines (and corresponding stock move) from a picking to another.
+        If wave is passed, attach new picking into it.
+        If not, attach line to their original picking.
+        :param stock.picking.batch wave: wave picking on which to put the move lines.
+        :param string description: description to use if no wave is given (in which case a new one is created.)
+        """
+        if not wave:
+            wave = self.env['stock.picking.batch'].create({
+                'is_wave': True,
+                'picking_type_id': self.picking_type_id and self.picking_type_id[0].id,
+                'user_id': self.env.context.get('active_owner_id'),
+                'description': description,
+            })
+            notification_title = _('The following wave transfer has been created')
+        else:
+            notification_title = _('The following wave transfer has been updated')
+        line_by_picking = defaultdict(lambda: self.env['stock.move.line'])
+        for line in self:
+            line_by_picking[line.picking_id] |= line
+        picking_to_wave_vals_list = []
+        split_pickings_ids = set()
+        for picking, lines in line_by_picking.items():
+            # Move the entire picking if all the line are taken.
+            line_by_move = defaultdict(lambda: self.env['stock.move.line'])
+            qty_by_move = defaultdict(float)
+            for line in lines:
+                move = line.move_id
+                line_by_move[move] |= line
+                qty = line.uom_id._compute_quantity(line.quantity, line.product_id.uom_id, rounding_method='HALF-UP')
+                qty_by_move[line.move_id] += qty
+
+            # If all moves are to be transferred to the wave, link the picking to the wave.
+            if lines == picking.move_line_ids and lines.move_id == picking.move_ids:
+                add_all_moves = True
+                for move, qty in qty_by_move.items():
+                    if move.uom_id.is_zero(qty):
+                        add_all_moves = False
+                        break
+                if add_all_moves:
+                    wave.picking_ids = [Command.link(picking.id)]
+                    continue
+
+            # Split the picking in two part to extract only line that are taken on the wave.
+            picking_to_wave_vals = picking.copy_data({
+                'move_ids': [],
+                'move_line_ids': [],
+                'batch_id': wave.id,
+                'scheduled_date': picking.scheduled_date,
+            })[0]
+            split_pickings_ids.add(picking.id)
+            for move, move_lines in line_by_move.items():
+                picking_to_wave_vals['move_line_ids'] += [Command.link(line.id) for line in lines]
+                # If all the lines of a stock move are taken, we change the picking on the move.
+                if move_lines == move.move_line_ids:
+                    picking_to_wave_vals['move_ids'] += [Command.link(move.id)]
+                    continue
+                # Otherwise, split the move.
+                qty = qty_by_move[move]
+                new_move = move._split(qty)
+                new_move[0]['move_line_ids'] = [Command.set(move_lines.ids)]
+                picking_to_wave_vals['move_ids'] += [Command.create(new_move[0])]
+
+            picking_to_wave_vals_list.append(picking_to_wave_vals)
+
+        if picking_to_wave_vals_list:
+            split_pickings = self.env['stock.picking'].browse(split_pickings_ids) | self.env['stock.picking'].create(picking_to_wave_vals_list)
+            split_pickings._add_to_wave_post_picking_split_hook()
+        if wave.picking_type_id.batch_auto_confirm:
+            wave.action_confirm()
+        if self.env.context.get('from_wave_form'):
+            return {'type': 'ir.actions.client', 'tag': 'soft_reload'}
+        return {
+            'type': 'ir.actions.client',
+            'tag': 'display_notification',
+            'params': {
+                'title': notification_title,
+                'message': '%s',
+                'links': [{
+                    'label': wave.name,
+                    'url': f'/odoo/wave-transfers/{wave.id}',
+                }],
+                'sticky': False,
+                'next': {'type': 'ir.actions.act_window_close'},
+            }
+        }
+
+    def _is_auto_waveable(self):
+        self.ensure_one()
+        if not self.picking_id \
+           or (self.picking_id.state != 'assigned' or self.uom_id.is_zero(self.quantity)) and not self.env.context.get('skip_auto_waveable')  \
+           or self.batch_id.is_wave \
+           or not self.picking_type_id._is_auto_wave_grouped() \
+           or (self.picking_type_id.wave_group_by_category and self.product_id.categ_id not in self.picking_type_id.wave_category_ids):  # noqa: SIM103
+            return False
+        return True
+
+    def _auto_wave(self):
+        """ Try to find compatible waves to attach the move lines to,
+        otherwise create new waves when possible/appropriate. """
+        batchable_line_ids = OrderedSet()
+        lines_nearest_parent_locations = defaultdict(lambda: self.env['stock.location'])
+        wave_locs_by_picking_type = {}
+        for picking_type in self.picking_type_id:
+            if not picking_type.wave_group_by_location:
+                continue
+            wave_locs_by_picking_type[picking_type] = set(picking_type.wave_location_ids.ids)
+
+        for line in self:
+            if not line._is_auto_waveable():
+                continue
+            if not line.picking_type_id.wave_group_by_location:
+                batchable_line_ids.add(line.id)
+                continue
+            # We want to find the most descendant location in the wave locations list that is a parent of the line location.
+            # Since the wave locations are ordered by complete_name (from the most descendant to the most ancestor), we can iterate in reverse order.
+            wave_locs_set = wave_locs_by_picking_type[line.picking_type_id]
+            loc = line.location_id
+            while (loc):
+                if loc.id in wave_locs_set:
+                    lines_nearest_parent_locations[line] = loc
+                    batchable_line_ids.add(line.id)
+                    break
+                loc = loc.location_id
+        batchable_lines = self.env['stock.move.line'].browse(batchable_line_ids)
+
+        remaining_line_ids = batchable_lines._auto_wave_lines_into_existing_waves(nearest_parent_locations=lines_nearest_parent_locations)
+        remaining_lines = self.env['stock.move.line'].browse(remaining_line_ids)
+        if remaining_lines:
+            remaining_lines._auto_wave_lines_into_new_waves(nearest_parent_locations=lines_nearest_parent_locations)
+
+    def _get_potential_existing_waves_extra_domain(self, domain_list, picking_type):
+        """Extend extra conditions here"""
+        return domain_list
+
+    def _get_potential_new_waves_extra_domain(self, domain_list, picking_type):
+        """Extend extra conditions here"""
+        return domain_list
+
+    def _is_potential_existing_wave_extra(self, wave):
+        """Extend extra conditions here"""
+        return True
+
+    def _is_new_potential_line_extra(self, potential_line):
+        """Extend extra conditions here"""
+        return True
+
+    def _auto_wave_lines_into_existing_waves(self, nearest_parent_locations=False):
+        """ Try to add move lines to existing waves if possible,
+        return move lines of which no appropriate waves were found to link to.
+
+        :param defaultdict nearest_parent_locations: the key is the move line
+            and the value is the nearest parent location in the wave locations list.
+        """
+        remaining_lines = OrderedSet()
+        batches_to_validate_ids = self.env.context.get('batches_to_validate', False)
+        for (picking_type, lines) in self.grouped(lambda l: l.picking_type_id).items():
+            if lines:
+                domains = [
+                    Domain('picking_type_id', '=', picking_type.id),
+                    Domain('company_id', 'in', lines.mapped('company_id').ids),
+                    Domain('is_wave', '=', True),
+                ]
+                if picking_type.batch_auto_confirm:
+                    domains.append(Domain('state', 'not in', ['done', 'cancel']))
+                else:
+                    domains.append(Domain('state', '=', 'draft'))
+                if picking_type.batch_group_by_partner:
+                    domains.append(Domain('picking_ids.partner_id', 'in', lines.move_id.partner_id.ids))
+                if picking_type.batch_group_by_destination:
+                    domains.append(Domain('picking_ids.partner_id.country_id', 'in', lines.move_id.partner_id.country_id.ids))
+                if picking_type.batch_group_by_src_loc:
+                    domains.append(Domain('picking_ids.location_id', 'in', lines.location_id.ids))
+                if picking_type.batch_group_by_dest_loc:
+                    domains.append(Domain('picking_ids.location_dest_id', 'in', lines.location_dest_id.ids))
+                if batches_to_validate_ids:
+                    domains.append(Domain('id', 'not in', batches_to_validate_ids))
+                domains = lines._get_potential_existing_waves_extra_domain(domains, picking_type)
+
+                potential_waves = self.env['stock.picking.batch'].search(Domain.AND(domains))
+                wave_to_new_lines = defaultdict(set)
+
+                # These dictionaries are used to enforce batch max lines/transfers/weight limits,
+                # each time a line is matched to a wave, we update the corresponding values.
+                wave_to_new_moves = defaultdict(set)
+                waves_to_new_pickings = defaultdict(set)
+                waves_new_extra_weight = defaultdict(float)
+
+                waves_nearest_parent_locations = defaultdict(int)
+                if picking_type.wave_group_by_location:
+                    valid_wave_ids = set()
+                    # We want to find the most descendant location in the wave
+                    # locations list that is a parent of all the lines in each wave.
+                    # We also want to exclude waves that have lines that are not in these locations.
+                    for wave in potential_waves:
+                        for wave_location in reversed(picking_type.wave_location_ids):
+                            if all(loc._child_of(wave_location) for loc in wave.move_line_ids.location_id):
+                                waves_nearest_parent_locations[wave] = wave_location.id
+                                valid_wave_ids.add(wave.id)
+                                break
+                    potential_waves = self.env['stock.picking.batch'].browse(valid_wave_ids)
+
+                for line in lines:
+                    wave_found = False
+                    for wave in potential_waves:
+                        if line.company_id != wave.company_id \
+                        or (picking_type.batch_group_by_partner and line.move_id.partner_id != wave.picking_ids.partner_id) \
+                        or (picking_type.batch_group_by_destination and line.move_id.partner_id.country_id != wave.picking_ids.partner_id.country_id) \
+                        or (picking_type.batch_group_by_src_loc and line.location_id != wave.picking_ids.location_id) \
+                        or (picking_type.batch_group_by_dest_loc and line.location_dest_id != wave.picking_ids.location_dest_id) \
+                        or (picking_type.wave_group_by_product and line.product_id != wave.move_line_ids.product_id) \
+                        or (picking_type.wave_group_by_category and line.product_id.categ_id != wave.move_line_ids.product_id.categ_id) \
+                        or (picking_type.wave_group_by_location and waves_nearest_parent_locations[wave] != nearest_parent_locations[line].id) \
+                        or not line._is_potential_existing_wave_extra(wave):
+                            continue
+
+                        wave_new_move_ids = wave_to_new_moves[wave]
+                        wave_new_picking_ids = waves_to_new_pickings[wave]
+                        wave_move_ids = set(wave.move_line_ids.mapped('move_id.id'))
+                        wave_picking_ids = set(wave.move_line_ids.mapped('picking_id.id'))
+                        # `is_line_auto_mergeable` is a method that checks if the line can be added to the wave without exceeding the limits
+                        # It takes as arguments the number of new moves that will be added to the wave, the number of new pickings that will be added to the wave
+                        # and the extra weight that will be added to the wave. So we need to check that the move/picking of the line is not already in the wave
+                        # so that we don't count them as new moves/pickings.
+                        if not wave._is_line_auto_mergeable(
+                            line.move_id.id not in wave_move_ids and line.move_id.id not in wave_new_move_ids and len(wave_new_move_ids) + 1,
+                            line.picking_id.id not in wave_picking_ids and line.picking_id.id not in wave_new_picking_ids and len(wave_new_picking_ids) + 1,
+                            waves_new_extra_weight[wave] + line.product_id.weight * line.quantity_product_uom
+                        ):
+                            continue
+
+                        if line.move_id.id not in wave_move_ids:
+                            wave_to_new_moves[wave].add(line.move_id.id)
+                        if line.picking_id.id not in wave_picking_ids:
+                            waves_to_new_pickings[wave].add(line.picking_id.id)
+                        waves_new_extra_weight[wave] += line.product_id.weight * line.quantity_product_uom
+                        wave_to_new_lines[wave].add(line.id)
+                        wave_found = True
+                        break
+                    if not wave_found:
+                        remaining_lines.add(line.id)
+                for wave, line_ids in wave_to_new_lines.items():
+                    lines = self.env['stock.move.line'].browse(line_ids)
+                    lines._add_to_wave(wave)
+        return list(remaining_lines)
+
+    def _auto_wave_lines_into_new_waves(self, nearest_parent_locations=False):
+        """ Create new waves for the move lines that could not be added to existing waves. """
+        picking_types = self.picking_type_id
+        for picking_type in picking_types:
+            lines = self.filtered(lambda l: l.picking_type_id == picking_type)
+            domains = [Domain([
+                ('id', 'in', lines.ids),
+                ('company_id', 'in', self.company_id.ids),
+                ('picking_id.state', '=', 'assigned'),
+                ('picking_type_id', '=', picking_type.id),
+                '|',
+                ('batch_id', '=', False),
+                ('batch_id.is_wave', '=', False),
+            ])]
+            if picking_type.batch_group_by_partner:
+                domains.append(Domain('move_id.partner_id', 'in', lines.move_id.partner_id.ids))
+            if picking_type.batch_group_by_destination:
+                domains.append(Domain('move_id.partner_id.country_id', 'in', lines.move_id.partner_id.country_id.ids))
+            if picking_type.batch_group_by_src_loc:
+                domains.append(Domain('location_id', 'in', lines.location_id.ids))
+            if picking_type.batch_group_by_dest_loc:
+                domains.append(Domain('location_dest_id', 'in', lines.location_dest_id.ids))
+            if picking_type.wave_group_by_product:
+                domains.append(Domain('product_id', 'in', lines.product_id.ids))
+            if picking_type.wave_group_by_category:
+                domains.append(Domain('product_id.categ_id', 'in', lines.product_id.categ_id.ids))
+            if picking_type.wave_group_by_location:
+                domains.append(Domain('location_id', 'child_of', picking_type.wave_location_ids.ids))
+            domains = lines._get_potential_new_waves_extra_domain(domains, picking_type)
+
+            potential_lines = self.env['stock.move.line'].search(Domain.AND(domains))
+            lines_nearest_parent_locations = defaultdict(int)
+            if picking_type.wave_group_by_location:
+                for line in potential_lines:
+                    for location in reversed(picking_type.wave_location_ids):
+                        if line.location_id._child_of(location):
+                            lines_nearest_parent_locations[line] = location.id
+                            break
+
+            line_to_lines = defaultdict(set)
+            matched_lines = set()
+            remaining_line_ids = OrderedSet()
+            for line in lines:
+                lines_found = False
+                if line.id in matched_lines:
+                    continue
+                for potential_line in potential_lines:
+                    if line.id == potential_line.id \
+                    or line.company_id != potential_line.company_id \
+                    or (picking_type.batch_group_by_partner and line.move_id.partner_id != potential_line.move_id.partner_id) \
+                    or (picking_type.batch_group_by_destination and line.move_id.partner_id.country_id != potential_line.move_id.partner_id.country_id) \
+                    or (picking_type.batch_group_by_src_loc and line.location_id != potential_line.location_id) \
+                    or (picking_type.batch_group_by_dest_loc and line.location_dest_id != potential_line.location_dest_id) \
+                    or (picking_type.wave_group_by_product and line.product_id != potential_line.product_id) \
+                    or (picking_type.wave_group_by_category and line.product_id.categ_id != potential_line.product_id.categ_id) \
+                    or (picking_type.wave_group_by_location and lines_nearest_parent_locations[potential_line] != nearest_parent_locations[line].id)  \
+                    or not line._is_new_potential_line_extra(potential_line):
+                        continue
+
+                    line_to_lines[line].add(potential_line.id)
+                    matched_lines.add(potential_line.id)
+                    lines_found = True
+                if not lines_found:
+                    remaining_line_ids.add(line.id)
+
+            for line, potential_line_ids in line_to_lines.items():
+                if line.batch_id.is_wave:
+                    continue
+
+                potential_lines = self.env['stock.move.line'].browse(potential_line_ids | {line.id})
+
+                # We want to make sure that batch/wave limits specified in the picking type are respected.
+                # We want also to reduce picking splits as much as possible. So we try to group as much as possible by sorting the lines by picking and move.
+                potential_lines = potential_lines.sorted(key=lambda l: (l.picking_id.id, l.move_id.id))
+
+                while potential_lines:
+                    new_wave = self.env['stock.picking.batch'].create({
+                        'is_wave': True,
+                        'picking_type_id': picking_type.id,
+                        'description': line._get_auto_wave_description(nearest_parent_locations[line]),
+                    })
+                    wave_move_ids = set()
+                    wave_picking_ids = set()
+                    wave_weight = 0
+
+                    wave_line_ids = set()
+
+                    for potential_line in potential_lines:
+                        if potential_line.batch_id.is_wave:
+                            continue
+                        wave_move_ids.add(potential_line.move_id.id)
+                        wave_picking_ids.add(potential_line.picking_id.id)
+                        wave_weight += potential_line.product_id.weight * potential_line.quantity_product_uom
+                        if new_wave._is_line_auto_mergeable(
+                            len(wave_move_ids),
+                            len(wave_picking_ids),
+                            wave_weight
+                        ):
+                            wave_line_ids.add(potential_line.id)
+                        else:
+                            break
+                    wave_lines = self.env['stock.move.line'].browse(wave_line_ids)
+                    wave_lines._add_to_wave(new_wave)
+                    potential_lines -= wave_lines
+
+            remaining_lines = self.env['stock.move.line'].browse(remaining_line_ids)
+            remaining_waves = self.env['stock.picking.batch'].create([{
+                'is_wave': True,
+                'picking_type_id': picking_type.id,
+                'description': remaining_line._get_auto_wave_description(nearest_parent_locations[remaining_line]),
+            } for remaining_line in remaining_lines])
+            for (line, wave) in zip(remaining_lines, remaining_waves):
+                line._add_to_wave(wave)
+
+    def _get_auto_wave_description(self, nearest_parent_location=False):
+        self.ensure_one()
+        description = self.picking_id._get_auto_batch_description()
+        description_items = []
+        if description:
+            description_items.append(description)
+
+        if self.picking_type_id.wave_group_by_product:
+            description_items.append(self.product_id.display_name)
+        if self.picking_type_id.wave_group_by_category:
+            description_items.append(self.product_id.categ_id.complete_name)
+        if self.picking_type_id.wave_group_by_location:
+            description_items.append(nearest_parent_location.complete_name)
+
+        description = ', '.join(description_items)
+        return description

--- a/addons/stock/models/stock_picking.py
+++ b/addons/stock/models/stock_picking.py
@@ -151,6 +151,12 @@ class StockPicking(models.Model):
     shipping_volume = fields.Float(
         "Volume for Shipping", compute="_compute_shipping_volume")
 
+    batch_id = fields.Many2one(
+        'stock.picking.batch', string='Batch Transfer',
+        help='Batch associated to this transfer',
+        check_company=True, index=True, copy=False)
+    batch_sequence = fields.Integer(string='Sequence')
+
     # Used to search on pickings
     product_id = fields.Many2one('product.product', 'Product', related='move_ids.product_id', readonly=True)
     lot_id = fields.Many2one('stock.lot', 'Lot/Serial Number', related='move_line_ids.lot_id', readonly=True)
@@ -191,7 +197,8 @@ class StockPicking(models.Model):
         'Reference must be unique per company!',
     )
 
-    @api.depends_context('formatted_display_name')
+    @api.depends('partner_id')
+    @api.depends_context('display_name_partner', 'formatted_display_name')
     def _compute_display_name(self):
         super()._compute_display_name()
         if self.env.context.get('formatted_display_name'):
@@ -203,6 +210,11 @@ class StockPicking(models.Model):
                     extra_display_info.append(picking.partner_id.display_name)
                 if extra_display_info:
                     picking.display_name += f'\t--{" ".join(extra_display_info)}--'
+        elif self.env.context.get('display_name_partner'):
+            for picking in self:
+                if not picking.partner_id:
+                    continue
+                picking.display_name = f"{picking.display_name} - {picking.partner_id.display_name}"
 
     def _compute_has_tracking(self):
         for picking in self:
@@ -639,11 +651,19 @@ class StockPicking(models.Model):
                 picking.with_context(mail_notrack=True).write({'scheduled_date': scheduled_date})
         pickings._autoconfirm_picking()
 
+        for picking, vals in zip(pickings, vals_list):
+            if vals.get('batch_id'):
+                if not picking.batch_id.picking_type_id:
+                    picking.batch_id.picking_type_id = picking.picking_type_id[0]
+                picking.batch_id._sanity_check()
+
         return pickings
 
     def write(self, vals):
         if vals.get('picking_type_id') and any(picking.state in ('done', 'cancel') for picking in self):
             raise UserError(_("Changing the operation type of this record is forbidden at this point."))
+
+        old_batches = self.batch_id
         if vals.get('picking_type_id'):
             picking_type = self.env['stock.picking.type'].browse(vals.get('picking_type_id'))
             for picking in self:
@@ -652,6 +672,14 @@ class StockPicking(models.Model):
                     vals['location_id'] = picking_type.default_location_src_id.id
                     vals['location_dest_id'] = picking_type.default_location_dest_id.id
         res = super().write(vals)
+
+        if vals.get('batch_id'):
+            old_batches.filtered(lambda b: not b.picking_ids).state = 'cancel'
+            if not self.batch_id.picking_type_id:
+                self.batch_id.picking_type_id = self.picking_type_id[0]
+            self.batch_id._sanity_check()
+            # assign batch users to batch pickings
+            self.batch_id.picking_ids.assign_batch_user(self.batch_id.user_id.id)
         if vals.get('date_done'):
             self.filtered(lambda p: p.state == 'done').move_ids.date = vals['date_done']
         if vals.get('signature'):
@@ -694,6 +722,28 @@ class StockPicking(models.Model):
         self.ensure_one()
         return self.picking_type_code == 'outgoing'
 
+    def action_add_operations(self):
+        view = self.env.ref('stock.view_move_line_tree_detailed')
+        # TODO: would be better to have an actual record for the action.
+        return {
+            'name': _('Add Operations'),
+            'type': 'ir.actions.act_window',
+            'view_mode': 'list',
+            'view': view,
+            'views': [(view.id, 'list')],
+            'res_model': 'stock.move.line',
+            'target': 'new',
+            'domain': [
+                ('picking_id', 'in', self.ids),
+                ('state', '!=', 'done')
+            ],
+            'context': dict(
+                self.env.context,
+                picking_to_wave=self.ids,
+                active_wave_id=self.env.context.get('active_wave_id').id,
+                search_default_by_location=True,
+            )}
+
     def action_confirm(self):
         self._check_company()
         if not self.env.context.get('skip_zero_demand_check') and not modules.module.current_test:
@@ -707,6 +757,10 @@ class StockPicking(models.Model):
 
         # run scheduler for moves forecasted to not have enough in stock
         self.move_ids.filtered(lambda move: move.state not in ('draft', 'cancel', 'done'))._trigger_scheduler()
+
+        if self.env.user.has_group('stock.group_stock_picking_batch'):
+            for picking in self:
+                picking._find_auto_batch()
         return True
 
     def action_assign(self):
@@ -728,6 +782,10 @@ class StockPicking(models.Model):
         self.move_ids._action_cancel()
         self.write({'is_locked': True})
         self.filtered(lambda x: not x.move_ids).state = 'cancel'
+        # Remove the cancelled operation from its batch if the batch has other active picking(s).
+        for picking in self:
+            if picking.batch_id and any(picking.state != 'cancel' for picking in picking.batch_id.picking_ids):
+                picking.batch_id = None
         return True
 
     def action_return(self):
@@ -819,6 +877,15 @@ class StockPicking(models.Model):
             "domain": [('id', 'in', next_transfers.ids)],
         }
 
+    def action_view_batch(self):
+        self.ensure_one()
+        return {
+            'type': 'ir.actions.act_window',
+            'res_model': 'stock.picking.batch',
+            'res_id': self.batch_id.id,
+            'view_mode': 'form',
+        }
+
     def _prepare_return_move_default_values(self, move_id):
         # +--------------------------------------------------------------------------------------------------------+
         # |       picking_pick     <--Move Orig--    picking_pack     --Move Dest-->   picking_ship
@@ -989,6 +1056,8 @@ class StockPicking(models.Model):
 
     def _is_single_transfer(self):
         # Overriden for batches.
+        if self.env.user.has_group('stock.group_stock_picking_batch'):
+            return len(self) == 1 or len(self.batch_id) == 1
         return len(self) == 1
 
     def _check_entire_pack(self):
@@ -1115,6 +1184,33 @@ class StockPicking(models.Model):
             pickings_not_to_backorder.with_context(cancel_backorder=True)._action_done()
         if pickings_to_backorder:
             pickings_to_backorder.with_context(cancel_backorder=False)._action_done()
+
+        # Remove done pickings from their batch.
+        if any(picking.state == 'done' for picking in self):
+            to_assign_ids = set()
+            if self and self.env.context.get('pickings_to_detach'):
+                pickings_to_detach = self.env['stock.picking'].browse(self.env.context['pickings_to_detach'])
+                pickings_to_detach.batch_id = False
+                pickings_to_detach.move_ids.filtered(lambda m: not m.quantity).picked = False
+                to_assign_ids.update(self.env.context['pickings_to_detach'])
+
+            for picking in self:
+                if picking.state != 'done':
+                    continue
+                # Avoid inconsistencies in states of the same batch when validating a single picking in a batch.
+                if picking.batch_id and any(p.state != 'done' for p in picking.batch_id.picking_ids):
+                    picking.batch_id = None
+                # If backorder were made and auto-batch is enabled, seek a batch for each of them with the selected criterias.
+                to_assign_ids.update(picking.backorder_ids.ids)
+
+            # If batch setting is enabled, auto-batch done pickings' backorders.
+            if self.env.user.has_group('stock.group_stock_picking_batch'):
+                # To avoid inconsistencies, all incorrect pickings must be removed before assigning backorder pickings
+                assignable_pickings = self.env['stock.picking'].browse(to_assign_ids)
+                for picking in assignable_pickings:
+                    picking._find_auto_batch()
+                assignable_pickings.move_line_ids.with_context(skip_auto_waveable=True)._auto_wave()
+
         report_actions = self._get_autoprint_report_actions()
         another_action = False
         pickings_show_report = self.filtered(lambda p: p.picking_type_id.auto_show_allocation_report)
@@ -1212,6 +1308,11 @@ class StockPicking(models.Model):
 
     def _should_show_transfers(self):
         """Whether the different transfers should be displayed on the pre action done wizards."""
+        if self.batch_id:
+            picking_ids_to_detach = self.env.context.get('pickings_to_detach') or []
+            batch_pickings_count = len(self.batch_id.picking_ids) - len(picking_ids_to_detach)
+            if len(self.batch_id) == 1 and len(self) == batch_pickings_count:
+                return False
         return len(self) > 1
 
     def _should_ignore_backorders(self):
@@ -1299,7 +1400,11 @@ class StockPicking(models.Model):
         """
         backorders = self.env['stock.picking']
         bo_to_assign = self.env['stock.picking']
+        pickings_to_detach = self.env['stock.picking'].browse(self.env.context.get('pickings_to_detach'))
         for picking in self:
+            # Avoid inconsistencies in states of the same batch when validating a single picking in a batch.
+            if picking.batch_id and picking.state != 'done' and any(p not in self for p in picking.batch_id.picking_ids - pickings_to_detach):
+                picking.batch_id = None
             if backorder_moves:
                 moves_to_backorder = backorder_moves.filtered(lambda m: m.picking_id == picking)
             else:
@@ -1872,3 +1977,126 @@ class StockPicking(models.Model):
                 'is_entire_pack': True,
             })
         return move_line_vals
+
+    # ===== Batch business methods =====
+
+    def _add_to_wave_post_picking_split_hook(self):
+        # Hook meant to be overriden
+        pass
+
+    def assign_batch_user(self, user_id):
+        pickings = self.filtered(lambda p: p.user_id.id != user_id)
+        pickings.write({'user_id': user_id})
+        for pick in pickings:
+            if user_id:
+                log_message = _('Assigned to %s Responsible', pick.batch_id._get_html_link())
+            else:
+                log_message = _('Unassigned responsible from %s', pick.batch_id._get_html_link())
+            pick.message_post(body=log_message)
+
+    def _get_possible_pickings_domain(self):
+        self.ensure_one()
+        domain = [
+            ('id', '!=', self.id),
+            ('company_id', '=', self.company_id.id if self.company_id else False),
+            ('state', '=', 'assigned'),
+            ('picking_type_id', '=', self.picking_type_id.id),
+            ('batch_id', '=', False),
+        ]
+        if self.picking_type_id.batch_group_by_partner:
+            domain.append(('partner_id', '=', self.partner_id.id))
+        if self.picking_type_id.batch_group_by_destination:
+            domain.append(('partner_id.country_id', '=', self.partner_id.country_id.id))
+        if self.picking_type_id.batch_group_by_src_loc:
+            domain.append(('location_id', '=', self.location_id.id))
+        if self.picking_type_id.batch_group_by_dest_loc:
+            domain.append(('location_dest_id', '=', self.location_dest_id.id))
+
+        return Domain(domain)
+
+    def _get_possible_batches_domain(self):
+        self.ensure_one()
+        domain = [
+            ('state', 'in', ('draft', 'in_progress') if self.picking_type_id.batch_auto_confirm else ('draft',)),
+            ('picking_type_id', '=', self.picking_type_id.id),
+            ('company_id', '=', self.company_id.id if self.company_id else False),
+            ('is_wave', '=', False)
+        ]
+        if self.picking_type_id.batch_group_by_partner:
+            domain.append(('picking_ids.partner_id', '=', self.partner_id.id))
+        if self.picking_type_id.batch_group_by_destination:
+            domain.append(('picking_ids.partner_id.country_id', '=', self.partner_id.country_id.id))
+        if self.picking_type_id.batch_group_by_src_loc:
+            domain.append(('picking_ids.location_id', '=', self.location_id.id))
+        if self.picking_type_id.batch_group_by_dest_loc:
+            domain.append(('picking_ids.location_dest_id', '=', self.location_dest_id.id))
+        if self.env.context.get('batches_to_validate'):
+            domain.append(('id', 'not in', self.env.context.get('batches_to_validate')))
+
+        return Domain(domain)
+
+    def _get_auto_batch_description(self):
+        """ Get the description of the automatically created batch based on the grouped pickings and grouping criteria """
+        self.ensure_one()
+        description_items = []
+        if self.picking_type_id.batch_group_by_partner and self.partner_id:
+            description_items.append(self.partner_id.name or '')
+        if self.picking_type_id.batch_group_by_destination and self.partner_id.country_id:
+            description_items.append(self.partner_id.country_id.name)
+        if self.picking_type_id.batch_group_by_src_loc and self.location_id:
+            description_items.append(self.location_id.display_name)
+        if self.picking_type_id.batch_group_by_dest_loc and self.location_dest_id:
+            description_items.append(self.location_dest_id.display_name)
+        return ', '.join(description_items)
+
+    def _find_auto_batch(self):
+        self.ensure_one()
+        # Check if auto_batch is enabled for this picking.
+        if not self.picking_type_id.auto_batch or\
+           not self.picking_type_id._is_auto_batch_grouped() or\
+           self.batch_id or not self.move_ids or\
+           not self._is_auto_batchable():
+            return False
+
+        # Try to find a compatible batch to insert the picking
+        possible_batches = self.env['stock.picking.batch'].sudo().search(self._get_possible_batches_domain())
+        for batch in possible_batches:
+            if batch._is_picking_auto_mergeable(self):
+                batch.picking_ids |= self
+                return batch
+
+        # If no batch were found, try to find a compatible picking and put them both in a new batch.
+        possible_pickings = self.env['stock.picking'].search(self._get_possible_pickings_domain())
+        new_batch_data = {
+            'picking_ids': [Command.link(self.id)],
+            'company_id': self.company_id.id if self.company_id else False,
+            'picking_type_id': self.picking_type_id.id,
+            'description': self._get_auto_batch_description()
+        }
+        for picking in possible_pickings:
+            if self._is_auto_batchable(picking):
+                # Add the picking to the new batch
+                new_batch_data['picking_ids'].append(Command.link(picking.id))
+                new_batch = self.env['stock.picking.batch'].sudo().create(new_batch_data)
+                if picking.picking_type_id.batch_auto_confirm:
+                    new_batch.action_confirm()
+                return new_batch
+
+        # If nothing was found after those two steps, then create a batch with the current picking alone
+        new_batch_data['user_id'] = self.user_id.id
+        new_batch = self.env['stock.picking.batch'].sudo().create(new_batch_data)
+        if self.picking_type_id.batch_auto_confirm:
+            new_batch.action_confirm()
+        return new_batch
+
+    def _is_auto_batchable(self, picking=None):
+        """ Verifies if a picking can be put in a batch with another picking without violating auto_batch constrains.
+        """
+        if self.state != 'assigned':
+            return False
+        if self.picking_type_id.batch_max_lines:
+            picking = picking or self.env['stock.picking']
+            moves_count = len(set(self.move_ids.ids + picking.move_ids.ids))
+            if moves_count > self.picking_type_id.batch_max_lines:
+                return False
+        return not self.picking_type_id.batch_max_pickings or self.picking_type_id.batch_max_pickings > 1

--- a/None
+++ b/addons/stock/models/stock_picking_batch.py
@@ -0,0 +1,481 @@
+# Part of Odoo. See LICENSE file for full copyright and licensing details.
+
+from markupsafe import Markup
+
+from odoo import api, fields, models, _
+from odoo.exceptions import UserError
+
+from odoo.addons.stock.models.stock_move import PROCUREMENT_PRIORITIES
+
+
+class StockPickingBatch(models.Model):
+    _name = 'stock.picking.batch'
+    _inherit = ['mail.thread', 'mail.activity.mixin']
+    _description = "Batch Transfer"
+    _order = "priority desc, name desc"
+
+    name = fields.Char(
+        string='Batch Transfer', default=lambda self: _('New'),
+        copy=False, required=True, readonly=True)
+    description = fields.Char('Description')
+    user_id = fields.Many2one(
+        'res.users', string='Responsible', tracking=True, check_company=True)
+    company_id = fields.Many2one(
+        'res.company', string="Company", required=True, readonly=True,
+        index=True, default=lambda self: self.env.company)
+    picking_ids = fields.One2many(
+        'stock.picking', 'batch_id', string='Transfers',
+        domain="[('id', 'in', allowed_picking_ids)]", check_company=True,
+        help='List of transfers associated to this batch')
+    show_check_availability = fields.Boolean(
+        compute='_compute_show_check_availability',
+        string='Show Reserve Button')
+    show_allocation = fields.Boolean(
+        compute='_compute_show_allocation',
+        string='Show Allocation Button')
+    allowed_picking_ids = fields.One2many('stock.picking', compute='_compute_allowed_picking_ids')
+    move_ids = fields.One2many(
+        'stock.move', string="Stock moves", compute='_compute_move_ids')
+    move_line_ids = fields.One2many(
+        'stock.move.line', string='Stock move lines',
+        compute='_compute_move_line_ids', inverse='_set_move_line_ids', search='_search_move_line_ids')
+    state = fields.Selection([
+        ('draft', 'Draft'),
+        ('in_progress', 'In progress'),
+        ('done', 'Done'),
+        ('cancel', 'Cancelled')], default='draft',
+        store=True, compute='_compute_state',
+        copy=False, tracking=True, required=True, readonly=True, index=True)
+    priority = fields.Selection(PROCUREMENT_PRIORITIES, string='Priority', default='0')
+    picking_type_id = fields.Many2one(
+        'stock.picking.type', 'Operation Type', check_company=True, copy=False,
+        index=True)
+    warehouse_id = fields.Many2one(
+        'stock.warehouse', related='picking_type_id.warehouse_id')
+    picking_type_code = fields.Selection(
+        related='picking_type_id.code')
+    scheduled_date = fields.Datetime(
+        'Scheduled Date', copy=False, store=True, readonly=False, compute="_compute_scheduled_date",
+        help="""Scheduled date for the transfers to be processed.
+              - If manually set then scheduled date for all transfers in batch will automatically update to this date.
+              - If not manually changed and transfers are added/removed/updated then this will be their earliest scheduled date
+                but this scheduled date will not be set for all transfers in batch.""")
+    is_wave = fields.Boolean('This batch is a wave')
+    show_lots_text = fields.Boolean(compute='_compute_show_lots_text')
+    estimated_shipping_weight = fields.Float(
+        "shipping_weight", compute='_compute_estimated_shipping_capacity', digits='Product Unit')
+    estimated_shipping_volume = fields.Float(
+        "shipping_volume", compute='_compute_estimated_shipping_capacity', digits='Product Unit')
+    properties = fields.Properties('Properties', definition='picking_type_id.batch_properties_definition', copy=True)
+
+    @api.depends('description')
+    @api.depends_context('add_to_existing_batch')
+    def _compute_display_name(self):
+        if not self.env.context.get('add_to_existing_batch'):
+            return super()._compute_display_name()
+        for batch in self:
+            batch.display_name = f"{batch.name}: {batch.description}" if batch.description else batch.name
+
+    @api.depends('picking_type_id')
+    def _compute_show_lots_text(self):
+        for batch in self:
+            batch.show_lots_text = batch.picking_ids and batch.picking_ids[0].show_lots_text
+
+    def _compute_estimated_shipping_capacity(self):
+        for batch in self:
+            estimated_shipping_weight = 0
+            estimated_shipping_volume = 0
+            done_package_ids = set()
+            # packs
+            for pack in batch.move_line_ids.result_package_id:
+                p_type = pack.package_type_id
+                if pack.shipping_weight:
+                    # shipping_weight was computed, so base_weight should be included.
+                    estimated_shipping_weight += pack.shipping_weight
+                    done_package_ids.add(pack.id)
+                elif p_type:
+                    estimated_shipping_weight += p_type.base_weight or 0
+                    estimated_shipping_volume += (p_type.packaging_length * p_type.width * p_type.height) / 1000.0**3
+            # move without packs
+            for move_line in batch.picking_ids.move_ids.move_line_ids:
+                if move_line.result_package_id.id in done_package_ids:
+                    continue
+                estimated_shipping_weight += move_line.product_id.weight * move_line.quantity_product_uom
+                estimated_shipping_volume += move_line.product_id.volume * move_line.quantity_product_uom
+            batch.estimated_shipping_weight = estimated_shipping_weight
+            batch.estimated_shipping_volume = estimated_shipping_volume
+
+    @api.depends('company_id', 'picking_type_id', 'state')
+    def _compute_allowed_picking_ids(self):
+        allowed_picking_states = ['waiting', 'confirmed', 'assigned']
+
+        for batch in self:
+            domain_states = list(allowed_picking_states)
+            # Allows to add draft pickings only if batch is in draft as well.
+            if batch.state == 'draft':
+                domain_states.append('draft')
+            domain = [
+                ('company_id', '=', batch.company_id.id),
+                ('state', 'in', domain_states),
+            ]
+            if batch.picking_type_id:
+                domain += [('picking_type_id', '=', batch.picking_type_id.id)]
+            batch.allowed_picking_ids = self.env['stock.picking'].search(domain)
+
+    @api.depends('picking_ids', 'picking_ids.move_ids')
+    def _compute_move_ids(self):
+        for batch in self:
+            batch.move_ids = batch.picking_ids.move_ids
+
+    @api.depends('picking_ids.show_check_availability')
+    def _compute_show_check_availability(self):
+        for batch in self:
+            batch.show_check_availability = any(picking.show_check_availability for picking in batch.picking_ids)
+
+    @api.depends('picking_ids', 'picking_ids.move_line_ids')
+    def _compute_move_line_ids(self):
+        for batch in self:
+            batch.move_line_ids = batch.picking_ids.move_line_ids
+
+    def _search_move_line_ids(self, operator, value):
+        return [('picking_ids.move_line_ids', operator, value)]
+
+    @api.depends('state', 'move_ids', 'picking_type_id')
+    def _compute_show_allocation(self):
+        self.show_allocation = False
+        for batch in self:
+            if batch.picking_type_id.auto_show_allocation_report:
+                batch.show_allocation = batch.picking_ids._get_show_allocation(batch.picking_type_id)
+
+    @api.depends('picking_ids', 'picking_ids.state')
+    def _compute_state(self):
+        batchs = self.filtered(lambda batch: batch.state not in ['cancel', 'done'])
+        for batch in batchs:
+            if not batch.picking_ids:
+                continue
+            # Cancels automatically the batch picking if all its transfers are cancelled.
+            if all(picking.state == 'cancel' for picking in batch.picking_ids):
+                batch.state = 'cancel'
+            # Batch picking is marked as done if all its not canceled transfers are done.
+            elif all(picking.state in ['cancel', 'done'] for picking in batch.picking_ids):
+                batch.state = 'done'
+
+    @api.depends('picking_ids', 'picking_ids.scheduled_date')
+    def _compute_scheduled_date(self):
+        for rec in self:
+            rec.scheduled_date = min(rec.picking_ids.filtered('scheduled_date').mapped('scheduled_date'), default=False)
+
+    @api.onchange('scheduled_date')
+    def onchange_scheduled_date(self):
+        if self.scheduled_date:
+            self.picking_ids.scheduled_date = self.scheduled_date
+
+    def _set_move_line_ids(self):
+        new_move_lines = self[0].move_line_ids
+        for picking in self.picking_ids:
+            old_move_lines = picking.move_line_ids
+            picking.move_line_ids = new_move_lines.filtered(lambda ml: ml.picking_id.id == picking.id)
+            move_lines_to_unlink = old_move_lines - new_move_lines
+            if move_lines_to_unlink:
+                move_lines_to_unlink.unlink()
+
+    # -------------------------------------------------------------------------
+    # CRUD
+    # -------------------------------------------------------------------------
+    @api.model_create_multi
+    def create(self, vals_list):
+        for vals in vals_list:
+            if vals.get('name', '/') == '/':
+                company_id = vals.get('company_id', self.env.company.id)
+                picking_type = self.env['stock.picking.type'].browse(vals.get('picking_type_id'))
+                if picking_type:
+                    sequence_code = 'picking.wave' if vals.get('is_wave') else 'picking.batch'
+                    vals['name'] = self._prepare_name(picking_type, sequence_code, company_id)
+        return super().create(vals_list)
+
+    def write(self, vals):
+        batches_to_rename = self.env['stock.picking.batch']
+        if vals.get('picking_type_id'):
+            picking_type = self.env['stock.picking.type'].browse(vals.get('picking_type_id'))
+            batches_to_rename = self.filtered(lambda b: b.picking_type_id != picking_type)
+        res = super().write(vals)
+        if not self.picking_ids:
+            self.filtered(lambda b: b.state == 'in_progress').action_cancel()
+        if vals.get('picking_type_id'):
+            self._sanity_check()
+            for batch in batches_to_rename:
+                sequence_code = 'picking.wave' if batch.is_wave else 'picking.batch'
+                batch.name = self._prepare_name(picking_type, sequence_code, batch.company_id)
+        if vals.get('picking_ids'):
+            batch_without_picking_type = self.filtered(lambda batch: not batch.picking_type_id)
+            if batch_without_picking_type:
+                picking = self.picking_ids and self.picking_ids[0]
+                batch_without_picking_type.picking_type_id = picking.picking_type_id.id
+        if 'user_id' in vals:
+            self.picking_ids.assign_batch_user(vals['user_id'])
+        return res
+
+    @api.ondelete(at_uninstall=False)
+    def _unlink_if_not_done(self):
+        if any(batch.state == 'done' for batch in self):
+            raise UserError(_("You cannot delete Done batch transfers."))
+
+    # -------------------------------------------------------------------------
+    # Action methods
+    # -------------------------------------------------------------------------
+    def action_confirm(self):
+        """Sanity checks, confirm the pickings and mark the batch as confirmed."""
+        self.ensure_one()
+        if not self.picking_ids:
+            raise UserError(_("You have to set some pickings to batch."))
+        self.picking_ids.action_confirm()
+        self._check_company()
+        self.state = 'in_progress'
+        return True
+
+    def action_cancel(self):
+        self.state = 'cancel'
+        self.picking_ids = False
+        return True
+
+    def action_print(self):
+        self.ensure_one()
+        return self.env.ref('stock.action_report_picking_batch').report_action(self)
+
+    def action_done(self):
+        def has_no_quantity(picking):
+            return all(not m.picked or m.uom_id.is_zero(m.quantity) for m in picking.move_ids if m.state not in ('done', 'cancel'))
+
+        def is_empty(picking):
+            return all(m.uom_id.is_zero(m.quantity) for m in picking.move_ids if m.state not in ('done', 'cancel'))
+
+        self.ensure_one()
+        self._check_company()
+        # Empty 'assigned' or 'waiting for another operation' pickings will be removed from the batch when it is validated.
+        pickings = self.mapped('picking_ids').filtered(lambda picking: picking.state not in ('cancel', 'done'))
+        empty_waiting_pickings = self.mapped('picking_ids').filtered(lambda p: (p.state in ('waiting', 'confirmed') and has_no_quantity(p)) or (p.state == 'assigned' and is_empty(p)))
+        pickings = pickings - empty_waiting_pickings
+
+        empty_pickings = pickings.filtered(has_no_quantity)
+
+        # Run sanity_check as a batch and ignore the one in button_validate() since it is done here.
+        pickings._sanity_check(separate_pickings=False)
+        context = {
+            'skip_sanity_check': True,   # Skip sanity_check in pickings button_validate()
+            'pickings_to_detach': empty_waiting_pickings.ids,  # Remove 'waiting' pickings from the batch
+            'batches_to_validate': self.ids,  # Skip current batch in auto_wave
+        }
+        if len(empty_pickings) != len(pickings):
+            # If some pickings are at least partially done, other pickings (empty & waiting) will be removed from batch without being cancelled in case of no backorder
+            pickings = pickings - empty_pickings
+            context['pickings_to_detach'] = context['pickings_to_detach'] + empty_pickings.ids
+
+        for picking in pickings:
+            picking.message_post(
+                body=Markup("<b>%s:</b> %s <a href=#id=%s&view_type=form&model=stock.picking.batch>%s</a>") % (
+                    _("Transferred by"),
+                    _("Batch Transfer"),
+                    picking.batch_id.id,
+                    picking.batch_id.name))
+
+        if empty_waiting_pickings:
+            self.message_post(body=_(
+                "%s was removed from the batch, no quantity processed",
+                Markup(', ').join([picking._get_html_link() for picking in empty_waiting_pickings])
+            ))
+
+        return pickings.with_context(**context).button_validate()
+
+    def action_assign(self):
+        self.ensure_one()
+        self.picking_ids.action_assign()
+
+    def action_put_in_pack(self, *, package_id=False, package_type_id=False, package_name=False):
+        """ Action to put move lines with 'Done' quantities into a new pack
+        This method follows same logic to stock.picking.
+        """
+        self.ensure_one()
+        if self.state not in ('done', 'cancel'):
+            return self.move_line_ids.action_put_in_pack(package_id=package_id, package_type_id=package_type_id, package_name=package_name)
+
+    def action_view_allocation_report(self):
+        return self.env["ir.actions.actions"]._for_xml_id("stock.allocation_report_action")
+
+    def action_open_label_layout(self):
+        if self.env.user.has_group('stock.group_production_lot') and self.move_line_ids.lot_id:
+            view = self.env.ref('stock.picking_label_type_form')
+            return {
+                'name': _('Choose Type of Labels To Print'),
+                'type': 'ir.actions.act_window',
+                'res_model': 'picking.label.type',
+                'views': [(view.id, 'form')],
+                'target': 'new',
+                'context': {'default_picking_ids': self.picking_ids.ids},
+            }
+        return self.move_ids.action_open_label_layout()
+
+    def action_merge(self):
+        if not self:
+            return
+        if len(self) < 2:
+            raise UserError(self.env._('Please select at least two batch/wave transfers to merge.'))
+        if len(self.picking_type_id) > 1:
+            raise UserError(_('Batch/Wave transfers with different operation types cannot be merged.'))
+        if len(set(self.mapped('is_wave'))) > 1:
+            raise UserError(_('Batch transfers cannot be merged with wave transfers and vice versa.'))
+        if len(set(self.mapped('state'))) > 1:
+            raise UserError(_('Batch/Wave transfers with different states cannot be merged.'))
+        if self[0].state in ['done', 'cancel']:
+            raise UserError(_('You cannot merge done or cancelled batch/wave transfers.'))
+
+        target_batch = self[:1]
+        other_batches = self[1:]
+        earliest_batch = self.filtered(lambda b: b.scheduled_date).sorted(key=lambda b: b.scheduled_date)[0]
+        merged_batch_vals = earliest_batch._get_merged_batch_vals()
+        target_batch.move_line_ids |= other_batches.move_line_ids
+        target_batch.picking_ids |= other_batches.picking_ids
+        target_batch.write(merged_batch_vals)
+        other_batches.unlink()
+        return {
+            'type': 'ir.actions.client',
+            'tag': 'display_notification',
+            'params': {
+                'title': _('Batch/Wave transfers have been merged into the following transfer'),
+                'message': '%s',
+                'links': [{
+                    'label': target_batch.name,
+                    'url': f"/odoo/action-stock.{'action_picking_tree_wave' if target_batch.is_wave else 'stock_picking_batch_action'}/{target_batch.id}",
+                }],
+                'sticky': False,
+                'next': {'type': 'ir.actions.act_window_close'},
+            }
+        }
+
+    def action_batch_detailed_operations(self):
+        self.ensure_one()
+        view_id = self.env.ref('stock.stock_move_line_view_list_from_batch').id
+        return {
+            'name': _('Detailed Operations'),
+            'view_mode': 'list',
+            'type': 'ir.actions.act_window',
+            'res_model': 'stock.move.line',
+            'views': [(view_id, 'list')],
+            'domain': [('id', 'in', self.picking_ids.move_line_ids.ids)],
+            'context': {
+                'default_company_id': self.company_id.id,
+                'default_picking_id': self.picking_ids and self.picking_ids[0].id or False,
+                'picking_ids': self.picking_ids.ids,
+                'show_lots_text': self.show_lots_text,
+                'picking_code': self.picking_type_code,
+                'create': self.state not in ('done', 'cancel'),
+            }
+        }
+
+    def action_see_packages(self):
+        self.ensure_one()
+        if self.state == 'done':
+            return {
+                'name': self.env._("Packages"),
+                'res_model': 'stock.package.history',
+                'view_mode': 'list',
+                'views': [(False, 'list')],
+                'type': 'ir.actions.act_window',
+                'domain': [('picking_ids', 'in', self.picking_ids.ids)],
+                'context': {
+                    'search_default_main_packages': True,
+                }
+            }
+
+        return {
+            'name': self.env._("Packages"),
+            'res_model': 'stock.package',
+            'view_mode': 'list,kanban,form',
+            'views': [(self.env.ref('stock.stock_package_view_list_editable').id, 'list'), (False, 'kanban'), (False, 'form')],
+            'type': 'ir.actions.act_window',
+            'domain': [('picking_ids', 'in', self.picking_ids.ids)],
+            'context': {
+                'picking_ids': self.picking_ids.ids,
+                'location_id': self.picking_ids[:1].location_id.id,
+                'can_add_entire_packs': self.picking_type_code != 'incoming',
+                'search_default_main_packages': True,
+            },
+        }
+
+    # -------------------------------------------------------------------------
+    # Miscellaneous
+    # -------------------------------------------------------------------------
+    @api.model
+    def _prepare_name(self, picking_type, sequence_code, company_id):
+        sequence = self.env['ir.sequence'].with_company(company_id).next_by_code(sequence_code) or '/'
+        split = sequence.rsplit('/', 1)
+        if len(split) != 2:
+            if self:
+                self.message_post(
+                    body=_(
+                        "The sequence '%(sequence)s' is misconfigured. "
+                        "Its prefix should end with a '/' separator.",
+                        sequence=sequence_code,
+                    ),
+                )
+            return f"{picking_type.sequence_code}{sequence}"
+        sequence_prefix, sequence_number = split
+        return f"{sequence_prefix}/{picking_type.sequence_code}{sequence_number}"
+
+    def _sanity_check(self):
+        for batch in self:
+            if not batch.picking_ids <= batch.allowed_picking_ids:
+                erroneous_pickings = batch.picking_ids - batch.allowed_picking_ids
+                raise UserError(_(
+                    "The following transfers cannot be added to batch transfer %(batch)s. "
+                    "Please check their states and operation types.\n\n"
+                    "Incompatibilities: %(incompatible_transfers)s",
+                    batch=batch.name,
+                    incompatible_transfers=erroneous_pickings.mapped('name')))
+
+    def _track_log_get_default_subtype(self, track_init_values):
+        if 'state' in track_init_values:
+            return self.env.ref('stock.mt_batch_state')
+        return super()._track_log_get_default_subtype(track_init_values)
+
+    def _is_picking_auto_mergeable(self, picking):
+        """ Verifies if a picking can be safely inserted into the batch without violating auto_batch_constrains.
+        """
+        res = True
+        if self.picking_type_id.batch_max_lines:
+            res = res and (len(self.move_ids) + len(picking.move_ids) <= self.picking_type_id.batch_max_lines)
+        if self.picking_type_id.batch_max_pickings:
+            res = res and (len(self.picking_ids) + 1 <= self.picking_type_id.batch_max_pickings)
+        return res
+
+    def _is_line_auto_mergeable(self, num_of_moves=False, num_of_pickings=False, weight=False):
+        """ Verifies if a line can be safely inserted into the wave without violating auto_batch_constrains.
+        """
+        self.ensure_one()
+        res = True
+        if num_of_moves:
+            res = res and self._are_moves_auto_mergeable(num_of_moves)
+        if num_of_pickings:
+            res = res and self._are_pickings_auto_mergeable(num_of_pickings)
+        return res
+
+    def _are_moves_auto_mergeable(self, num_of_moves):
+        self.ensure_one()
+        res = True
+        if self.picking_type_id.batch_max_lines:
+            res = res and (len(self.move_ids) + num_of_moves <= self.picking_type_id.batch_max_lines)
+        return res
+
+    def _are_pickings_auto_mergeable(self, num_of_pickings):
+        self.ensure_one()
+        res = True
+        if self.picking_type_id.batch_max_pickings:
+            res = res and (len(self.picking_ids) + num_of_pickings <= self.picking_type_id.batch_max_pickings)
+        return res
+
+    def _get_merged_batch_vals(self):
+        self.ensure_one()
+        return {
+            'user_id': self.user_id.id,
+            'description': self.description,
+            'scheduled_date': self.scheduled_date,
+        }

--- a/addons/stock/models/stock_picking_type.py
+++ b/addons/stock/models/stock_picking_type.py
@@ -3,7 +3,7 @@ from ast import literal_eval
 from datetime import timedelta
 
 from odoo import _, api, fields, models
-from odoo.exceptions import UserError
+from odoo.exceptions import UserError, ValidationError
 from odoo.fields import Domain
 from odoo.tools import SQL
 from odoo.tools.translate import mark_as_copy
@@ -150,6 +150,49 @@ class StockPickingType(models.Model):
         default=lambda self: self.env.company.picking_policy, required=True,
         help="It specifies goods to be transferred partially or all at once")
 
+    # ==========================================================================
+    # Batch related fields
+    # ==========================================================================
+    count_picking_batch = fields.Integer(compute='_compute_picking_count')
+    count_picking_wave = fields.Integer(compute='_compute_picking_count')
+    auto_batch = fields.Boolean('Automatic Batches',
+        help="Automatically put pickings into batches as they are confirmed when possible.")
+    batch_group_by_partner = fields.Boolean('Contact', help="Automatically group batches by contacts.")
+    batch_group_by_destination = fields.Boolean('Destination Country', help="Automatically group batches by destination country.")
+    batch_group_by_src_loc = fields.Boolean('Group by Source Location',
+        help="Automatically group batches by their source location.")
+    batch_group_by_dest_loc = fields.Boolean('Group by Destination Location',
+        help="Automatically group batches by their destination location.")
+    wave_group_by_product = fields.Boolean('Product',
+        help="Split transfers by product then group transfers that have the same product.")
+    wave_group_by_category = fields.Boolean('Product Category',
+        help="Split transfers by product category, then group transfers that have the same product category.")
+    wave_category_ids = fields.Many2many('product.category',
+        string='Wave Product Categories', help="Categories to consider when grouping waves.")
+    wave_group_by_location = fields.Boolean('Location',
+        help="Split transfers by defined locations, then group transfers with the same location.")
+    wave_location_ids = fields.Many2many('stock.location',
+        string='Wave Locations',
+        help="Locations to consider when grouping waves.",
+        domain="[('usage', '=', 'internal')]")
+    batch_max_lines = fields.Integer("Maximum lines",
+        help="A transfer will not be automatically added to batches that will exceed this number of lines if the transfer is added to it.\n"
+             "Leave this value as '0' if no line limit.")
+    batch_max_pickings = fields.Integer("Maximum transfers",
+        help="A transfer will not be automatically added to batches that will exceed this number of transfers.\n"
+             "Leave this value as '0' if no transfer limit.")
+    batch_auto_confirm = fields.Boolean("Auto-confirm", default=True)
+    batch_properties_definition = fields.PropertiesDefinition('Batch Properties')
+
+    @api.constrains(lambda self: self._get_batch_group_by_keys() + ['auto_batch'])
+    def _check_auto_batch_options(self):
+        group_by_keys = self._get_batch_and_wave_group_by_keys()
+        for picking_type in self:
+            if not picking_type.auto_batch:
+                continue
+            if not any(picking_type[key] for key in group_by_keys):
+                raise ValidationError(_("If the Automatic Batches feature is enabled, at least one 'Group by' option must be selected."))
+
     @api.model_create_multi
     def create(self, vals_list):
         for vals in vals_list:
@@ -246,22 +289,45 @@ class StockPickingType(models.Model):
             rec.hide_reservation_method = rec.code == 'incoming'
 
     def _compute_picking_count(self):
+        base_domain = Domain([('state', 'not in', ('done', 'cancel')), ('picking_type_id', 'in', self.ids)])
+        # TODO (to check if it's not overkill): to avoid some `_read_group`:
+        #   - `count_picking` could be the sum of `count_picking_waiting` and `count_picking_ready`;
+        #   - We could compute some fields like we did for `count_picking_wave` and `count_picking_batch`
+        #     by checking a field value (would be easy for `state` and `backorder_id`.)
+        active_picking_states = ('assigned', 'waiting', 'confirmed')
         domains = {
-            'count_picking_draft': [('state', '=', 'draft')],
-            'count_picking_waiting': [('state', 'in', ('confirmed', 'waiting'))],
-            'count_picking_ready': [('state', '=', 'assigned')],
-            'count_picking': [('state', 'in', ('assigned', 'waiting', 'confirmed'))],
-            'count_picking_late': [('state', 'in', ('assigned', 'waiting', 'confirmed')), '|', ('scheduled_date', '<', fields.Date.today()), ('has_deadline_issue', '=', True)],
-            'count_picking_backorders': [('backorder_id', '!=', False), ('state', 'in', ('confirmed', 'assigned', 'waiting'))],
+            'count_picking_draft': Domain('state', '=', 'draft'),
+            'count_picking_waiting': Domain('state', 'in', ('confirmed', 'waiting')),
+            'count_picking_ready': Domain('state', '=', 'assigned'),
+            'count_picking': Domain('state', 'in', active_picking_states),
+            'count_picking_late': Domain([
+                ('state', 'in', active_picking_states),
+                '|',
+                ('scheduled_date', '<', fields.Date.today()),
+                ('has_deadline_issue', '=', True),
+            ]),
+            'count_picking_backorders': Domain([
+                ('backorder_id', '!=', False),
+                ('state', 'in', active_picking_states),
+            ]),
         }
         for field_name, domain in domains.items():
-            data = self.env['stock.picking']._read_group(domain +
-                [('state', 'not in', ('done', 'cancel')), ('picking_type_id', 'in', self.ids)],
-                ['picking_type_id'], ['__count'])
+            data = self.env['stock.picking']._read_group(
+                Domain.AND([base_domain, domain]), ['picking_type_id'], ['__count'])
             count = {picking_type.id: count for picking_type, count in data}
             for record in self:
                 record[field_name] = count.get(record.id, 0)
 
+        if self.env.user.has_group('stock.group_stock_picking_batch'):
+            data = self.env['stock.picking.batch']._read_group(
+                base_domain, ['picking_type_id', 'is_wave'], ['__count'])
+            count = {(picking_type.id, is_wave): count for picking_type, is_wave, count in data}
+            for record in self:
+                record.count_picking_wave = count.get((record.id, True), 0)
+                record.count_picking_batch = count.get((record.id, False), 0)
+        else:
+            self.count_picking_batch, self.count_picking_wave = 0, 0
+
     @api.depends('warehouse_id')
     def _compute_display_name(self):
         """ Display 'Warehouse_name: PickingType_name' """
@@ -400,6 +466,18 @@ class StockPickingType(models.Model):
         action["context"] = dict(literal_eval(action["context"]), search_default_name="Barcode")
         return action
 
+    def action_batch(self):
+        action = self._get_action('stock.stock_picking_batch_action')
+        if self.env.context.get("view_mode"):
+            del action["mobile_view_mode"]
+            del action["views"]
+            action["view_mode"] = self.env.context["view_mode"]
+        return action
+
+    def action_wave(self):
+        action = self._get_action('stock.action_picking_tree_wave')
+        return action
+
     def _get_action(self, action_xmlid):
         action = self.env["ir.actions.actions"]._for_xml_id(action_xmlid)
         context = {}
@@ -519,3 +597,25 @@ class StockPickingType(models.Model):
             'internal': _('Internal Move'),
         }
         return code_names.get(self.code)
+
+    # ===== Batch business methods =====
+
+    def _is_auto_batch_grouped(self):
+        self.ensure_one()
+        return self.auto_batch and any(self[key] for key in self._get_batch_group_by_keys())
+
+    def _is_auto_wave_grouped(self):
+        self.ensure_one()
+        return self.auto_batch and any(self[key] for key in self._get_wave_group_by_keys())
+
+    @api.model
+    def _get_batch_group_by_keys(self):
+        return ['batch_group_by_destination', 'batch_group_by_src_loc', 'batch_group_by_dest_loc', 'batch_group_by_partner']
+
+    @api.model
+    def _get_wave_group_by_keys(self):
+        return ['wave_group_by_category', 'wave_group_by_location', 'wave_group_by_product']
+
+    @api.model
+    def _get_batch_and_wave_group_by_keys(self):
+        return self._get_batch_group_by_keys() + self._get_wave_group_by_keys()

--- a/addons/stock/models/stock_warehouse.py
+++ b/addons/stock/models/stock_warehouse.py
@@ -989,6 +989,8 @@ class StockWarehouse(models.Model):
                 'sequence': max_sequence + 1,
                 'company_id': self.company_id.id,
                 'auto_show_allocation_report': True,
+                'auto_batch': self.env.user.has_group('stock.group_stock_picking_batch'),
+                'batch_group_by_partner': self.env.user.has_group('stock.group_stock_picking_batch'),
             }, 'out_type_id': {
                 'name': _('Delivery Orders'),
                 'code': 'outgoing',
@@ -996,6 +998,8 @@ class StockWarehouse(models.Model):
                 'sequence': max_sequence + 7,
                 'print_label': True,
                 'company_id': self.company_id.id,
+                'auto_batch': self.env.user.has_group('stock.group_stock_picking_batch'),
+                'batch_group_by_partner': self.env.user.has_group('stock.group_stock_picking_batch'),
             }, 'pack_type_id': {
                 'name': _('Pack'),
                 'code': 'internal',
